### PR TITLE
Asset Scales: Detect asset transfers in transactions and balance-check

### LINKED ISSUE #824: Transaction status & balances not correctly updating
### Discord Discussion Link

https://discord.com/channels/808358975287722045/888500174685614090/930611178160390184

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Send ETH to another EOA

### What did not work?

- Transaction completed but status in wallet still pending (0.04 ETH)
- Wallet balance not updated with this transaction (0.04 ETH)
- Wallet was able to create a send transaction for this 0.04 ETH, that was not in the wallets balance. Led to a failed transaction creation. https://etherscan.io/tx/0xe6001517b9c1a92400cbaffa926981b2c48a6acd0f6c77e39370980388327447

From the user: "Latest update, tried to send 0.02 eth and it worked. I got my 0.04 eth but it still says pending and $ balance always at 0. First send was rejected because I selected max amount and there probably were gas price changes and transaction failed. Send send went well and I got 0.02 eth. I still have 0.01 eth in my balance, but it shows $0 which is confusing. In conclusion, everything worked well in the background but wallet interface wasn't updating balance and events properly."

image

Other relevant Discord links:
https://discord.com/channels/808358975287722045/888500174685614090/930623737475248189
https://discord.com/channels/808358975287722045/888500174685614090/930632764095614978

### Version

Community Edition v0.10.1

### Relevant log output

_No response_

### LINKED ISSUE #907: DAI balance not displaying 20 mins after txn
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Withdraw DAI from my yearn vault and see it in my Tally wallet

### What did not work?

My yvDAI balance disappeared... but DAI never appeared under my assets. I can see DAI on etherscan, just not in tally https://etherscan.io/address/0x1cbb77b65b07403ad7f85ff073bfc29a8212219c#tokentxns 

Screen Shot 2022-01-27 at 6 06 09 PM

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #997: USDT balance after swap not showing
### Discord Discussion Link

https://discord.com/channels/808358975287722045/888500174685614090/943018041279647774

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Swap WETH to USDT.

Chrome Browser: Version 98.0.4758.80 (Official Build) (arm64)

### What did not work?

"Just did a successful swap of 0.01 WETH to USDT (~$29). Only problem is the USDT tokens don't show in my wallet.  Have a confirmation on Etherescan and my WETH updated properly."

Users has updated: This remedied itself. I removed and reinstalled the extension. --> The USDT balance is now showing

Similar issue reported by Nico, in this case USDT and MANA
https://discord.com/channels/808358975287722045/888500174685614090/942941150380191744

He hasn't reinstalled the wallet yet since it's a 24 word seed wallet.

### Version

Community Edition v0.11.2

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Shadowfiend**
Asset transfers for any known assets are now detected in new
transactions, and the associated assets are marked for fast
balance checks (within a minute).

To do this, a few things land:
- EVM transactions now carry their associated logs.
- The ERC20 helper lib allows parsing ERC20 transfer logs to
  well-typed objects.
- Transaction annotations now have a concept of subannotations,
  which are annotations for things that may happen internally in
  a transaction, vs the base annotation that annotates the nature
  of the transaction itself.
- The enrichment service registers ERC20 transfer logs for known
  assets in transactions as subannotations. 
- The indexing service is now notified of enriched transactions,
  and uses these to note if any internal asset transfers necessitate
  balance checks.

NOTE: this should apply not only to Tally swap transactions, but to
all transactions seen by the wallet.

Closes #997, closes #907, closes #824.

### CODE REVIEW

**Comment by Shadowfiend** [comment]
Gah, built this assuming #994 was merged for some reason 🙈

**Review by mhluongo — COMMENTED** [review]
Concept ACK, excited about this!

File: background/lib/erc20.ts [hunk 1]
```diff
@@ -1,33 +1,65 @@
...
-  "function decimals() view returns (uint8)",
-  "function name() view returns (string)",
-  "function symbol() view returns (string)",
-  "function totalSupply() view returns (uint256)",
-  "function transfer(address to, uint amount) returns (bool)",
-  "function transferFrom(address from, address to, uint amount) returns (bool)",
-  "event Transfer(address indexed from, address indexed to, uint amount)",
-  "event Approval(address indexed owner, address indexed spender, uint amount)",
-]
+const ERC20_FUNCTIONS = {
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply background/lib/erc20.ts:17]: I think I would lose the known types of keys unfortunately, which is part of my goal here.

[the remark this answers is not in the document]
Comment (Shadowfiend) [reply background/lib/erc20.ts:17]: Oh I see what you mean... Yeah that might work, will give it a shot.

[the remark this answers is not in the document]
Comment (Shadowfiend) [reply background/lib/erc20.ts:17]: Nah, I was right. Can't use `map` effectively without going through `Object.entries`, and `Object.entries` loses the const typing on the keys.

**Comment by Shadowfiend** [comment]
Should be good to go n

**Review by mhluongo — APPROVED** [review]
Local tests look good, code looks great.

**Pull request merged**